### PR TITLE
[CELEBORN-1203] Add `LicenseAndNoticeMergeStrategy` to resolve inner project LICENSE/NOTICE conflict for shaded client packaging

### PR DESCRIPTION
**Pull request by cfmcgrady**

### CODE REVIEW

File: project/CelebornBuild.scala [hunk 1]
```diff
@@ -753,7 +753,13 @@ trait SparkClientProjects {
   
         (assembly / assemblyMergeStrategy) := {
           case m if m.toLowerCase(Locale.ROOT).endsWith("manifest.mf") => MergeStrategy.discard
-          case "META-INF/LICENSE" | "META-INF/NOTICE" => MergeStrategy.preferProject
+          // the LicenseAndNoticeMergeStrategy always picks the license/notice file from the current project
+          case m @ ("META-INF/LICENSE" | "META-INF/NOTICE") => CustomMergeStrategy("LicenseAndNoticeMergeStrategy") { conflicts =>
+            val entry = conflicts.head
```
Comment (pan3793) [inline project/CelebornBuild.scala:758]: is this way reliable?

Comment (pan3793) [reply project/CelebornBuild.scala:758]: seems we can define a custom MergeStragegy class to eliminate the duplication

Comment (cfmcgrady) [reply project/CelebornBuild.scala:758]: > is this way reliable?
The strategy takes effect only when the conflict files are non-empty. The built-in strategy `MergeStrategy.first` encompasses an identical function call

https://github.com/sbt/sbt-assembly/blob/64dedc8f6f829fbe5feca9d59061a5e344419bd2/src/main/scala/sbtassembly/MergeStrategy.scala#L71-L73

Comment (cfmcgrady) [reply project/CelebornBuild.scala:758]: > seems we can define a custom MergeStragegy class to eliminate the duplication

It's hard to do, I need to conduct more in-depth research.

```
[error] `value` can only be used within a task or setting macro, such as :=, +=, ++=, Def.task, or Def.setting.
[error]     val projectLicenseFile = (Compile / resourceDirectory).value / entry.target
```

**Comment by cfmcgrady** [comment]
checked all the LICENSE/NOTICE files, it looks as expected after this PR.

File: project/CelebornBuild.scala (hunk 1 above)
Comment (pan3793) [reply project/CelebornBuild.scala:758]: I'm not sure if the `conflicts.head` always returns the current project file(if provided), but the test result looks good. please at least document something above `conflicts.head`

**Review by pan3793 — APPROVED** [review]

File: project/CelebornBuild.scala (hunk 1 above)
Comment (cfmcgrady) [reply project/CelebornBuild.scala:758]: > I'm not sure if the `conflicts.head` always returns the current project file(if provided)
no matter what the `conflicts.head` returns, the content of LICENSE/NOTICE will always be read from the current project resource directory.

**Comment by cfmcgrady** [comment]
thanks, merging to main(v0.5.0)/branch-0.4(v0.4.0).

**Pull request closed without merging**